README.md
<h2 align="center">
  <a href="https://smart-maas.eu/en/"><img src="https://github.com/SmartMaaS-Services/Transaction-Context-Manager/blob/main/docs/images/Header.jpeg" alt="Smart MaaS" width="500"></a>
  <br>
      SMART MOBILITY SERVICE PLATFORM
  <br>
  <a href="https://smart-maas.eu/en/"><img src="https://github.com/SmartMaaS-Services/Transaction-Context-Manager/blob/main/docs/images/Logos-Smart-MaaS.png" alt="Smart MaaS" width="250"></a>
  <br>
</h2>

<p align="center">
  <a href="mailto:info@smart-maas.eu">Contact</a> •
  <a href="https://github.com/SmartMaaS-Services/DigiTwin_APP/issues">Issues</a> •
  <a href="https://smart-maas.eu/en/">Project Page</a>
</p>


***

<h1 align="center">
  <a>
    DigiTwin APP
  </a>
</h1>

***


Die DigiTwin APP ist das Beispielhafte Frontend zum DigiTiwin.

Hier werden die Datenmodelle

* MobilityRegion
* MobilityStation
* MobilityService

Zur Anwendung gebracht.


Eine *MobilityRegion* ist ein zusammenhängender Verkehrsraum in dem ein hoher Anteil des täglichen Verkehrs abgewickelt wird.

Die Teilnehmer nutzen alle Möglichkeiten der Mobilität und werden innerhalb einer dieser MobilityRegion durch das Angebot von *MobilityStations* zum umstieg zwischen den Verkehrsmitteln ermutigt.
Ziel dieser Stationen ist es möglichst viel Individualverkehr in ÖPNV zu wandeln den Anreiz zu schaffen alternative Verkehrsmittel zu nutzen.

Im Rahmen von SmartMaaS wurde das Konzept dieser MobilityRegions auf eine FIWARE basierte Plattform übertragen.

Dazu wurden neue Datenmodelle geschaffen die untereinander die Struktur der MobilityRegion abbilden können.

![Uebersicht](/doc/images/Uebersicht_DigiTwin.png)



Eine Routing Funktionalität ist über TripGo in der DigiTwin App aufgenommen worden.

Die App liefert den Status der Mobilitätsstation übersichtlich als Sidebar.

Hier kat man je nach angebotenem MobilityService die Informationen zu z.B. den Parkplätzen, den Rädern der Sprottenflotte oder den Abfahrzeiten.

## Umsetzung

Aufbauend auf eine FIWARE Installation für die KielRegion wurde ein Tenant DigiTwin erstellt der zum einen die neuen Datenmodelle MobilityRegion, MobilityStation und MobilityService aufnimmt und zum anderen die vorhandenen Datenmodelle zu MobilityServices macht.
So ist die Verwaltung der App an diese 3 DatenModell Typen gebunden.

Es muss eine MobilityRegion erzeugt werden:

### MobilityRegion

	{
        "id": "urn:ngsi:MobilityRegion:KielRegion",
        "type": "MobilityRegion",
        "description": {
            "type": "Text",
            "value": "Mobilitätsplattform der KielRegion GmbH"
        },
        "location": {
            "type": "geo:json",
            "value": {
                "type": "Point",
                "coordinates": [
                    10.1329672,
                    54.3223492
                ]
            }
        },
        "name": {
            "type": "Text",
            "value": "KielRegion"
        }
	}

Dieser MobitityRegion werden dann die MobilityStations zugeordnet:

### MobitityRegion

	[
    {
        "id": "urn:ngsi:MobilityStation:KielRegion:Oppendorf",
        "type": "MobilityStation",
        "description": {
            "type": "Text",
            "value": "Mobilitätsstation Kiel - Bahnhof Oppendorf"
        },
        "location": {
            "type": "geo:json",
            "value": {
                "type": "Point",
                "coordinates": [
                    10.2086699,
                    54.3262973
                ]
            }
        },
        "name": {
            "type": "Text",
            "value": "Kiel - Bahnhof Oppendorf"
        },
        "refMobilityRegion": {
            "type": "Text",
            "value": "urn:ngsi:MobilityRegion:KielRegion"
        }
    },
    {
        "id": "urn:ngsi:MobilityStation:KielRegion:Hamdorf",
        "type": "MobilityStation",
        "description": {
            "type": "Text",
            "value": "Mobilitätsstation Kiel - Hamdorf",
            "metadata": {}
        },
        "location": {
            "type": "geo:json",
            "value": {
                "type": "Point",
                "coordinates": [
                    9.514357,
                    54.226997
                ]
            }
        },
        "name": {
            "type": "Text",
            "value": "Hamdorf"
        },
        "refMobilityRegion": {
            "type": "Text",
            "value": "urn:ngsi:MobilityRegion:KielRegion"
        }
    }
	]

Die Mobilitätsstation Russee ist noch nicht gebaut worden.

Darum haben wir nur Oppendorf als Beispiel und Hamdorf als eine zweite Station in den Daten.

Die an den Stationen angebotenen *MobilityServices* werden im 3. Datenmodell aufgenommen und können dort den Gegebenheiten der jeweiligen Station angepasst werden.

### MobilityServices

	[
    {
        "id": "urn:ngsi:MobilityService:KielRegion:Oppendorf:Parking",
        "type": "MobilityService",
        "description": {
            "type": "Text",
            "value": "Parkplatz Bahnhof Oppendorf"
        },
        "location": {
            "type": "geo:json",
            "value": {
                "type": "Point",
                "coordinates": [
                    10.208399,
                    54.326328
                ]
            }
        },
        "name": {
            "type": "Text",
            "value": "Parkplatz"
        },
        "refMobilityStation": {
            "type": "Text",
            "value": "urn:ngsi:MobilityStation:KielRegion:Oppendorf"
        },
        "refServiceProvider": {
            "type": "Text",
            "value": "urn:ngsi:OffStreetParking:KielRegion:Oppendorf"
        },
        "service": {
            "type": "Text",
            "value": "/parking/offstreet"
        }
    },
    {
        "id": "urn:ngsi:MobilityService:KielRegion:Oppendorf:BikeHire",
        "type": "MobilityService",
        "description": {
            "type": "Text",
            "value": "Sprottenflotte Bahnhof Oppendorf"
        },
        "location": {
            "type": "geo:json",
            "value": {
                "type": "Point",
                "coordinates": [
                    10.207911,
                    54.326055
                ]
            }
        },
        "name": {
            "type": "Text",
            "value": "Sprottenflotte"
        },
        "refMobilityStation": {
            "type": "Text",
            "value": "urn:ngsi:MobilityStation:KielRegion:Oppendorf"
        },
        "refServiceProvider": {
            "type": "Text",
            "value": "urn:ngsi:BikeHireDockingStation:KielRegion:20050"
        },
        "service": {
            "type": "Text",
            "value": "/transportation/bikehiredockingstations"
        }
    },
    {
        "id": "urn:ngsi:MobilityService:KielRegion:Oppendorf:Bus",
        "type": "MobilityService",
        "description": {
            "type": "Text",
            "value": "KVG Bushaltestelle Bahnhof Oppendorf"
        },
        "location": {
            "type": "geo:json",
            "value": {
                "type": "Point",
                "coordinates": [
                    10.208553,
                    54.326222
                ]
            }
        },
        "name": {
            "type": "Text",
            "value": "Bushaltestelle"
        },
        "refMobilityStation": {
            "type": "Text",
            "value": "urn:ngsi:MobilityStation:KielRegion:Oppendorf"
        },
        "refServiceProvider": {
            "type": "Text",
            "value": "urn:ngsi:TransportStation:KVG:284"
        },
        "service": {
            "type": "Text",
            "value": "/transportation/transportstations"
        }
    },
    {
        "id": "urn:ngsi:MobilityService:KielRegion:Oppendorf:Train",
        "type": "MobilityService",
        "description": {
            "type": "Text",
            "value": "DB Bahnstation Bahnhof Oppendorf"
        },
        "location": {
            "type": "StructuredValue",
            "value": {
                "type": "Point",
                "coordinates": [
                    10.208855,
                    54.326332
                ]
            }
        },
        "name": {
            "type": "Text",
            "value": "Bahnstation"
        },
        "refMobilityStation": {
            "type": "Text",
            "value": "urn:ngsi:MobilityStation:KielRegion:Oppendorf"
        },
        "refServiceProvider": {
            "type": "Text",
            "value": "urn:ngsi:TransportStation:DB:08004"
        },
        "service": {
            "type": "Text",
            "value": "/transportation/transportstations"
        }
    },
    {
        "id": "urn:ngsi:MobilityService:KielRegion:Oppendorf:AirQuality",
        "type": "MobilityService",
        "description": {
            "type": "Text",
            "value": "Infoservice Luftqualität Bahnhof Oppendorf"
        },
        "location": {
            "type": "geo:json",
            "value": {
                "type": "Point",
                "coordinates": [
                    10.208026,
                    54.326155
                ]
            }
        },
        "name": {
            "type": "Text",
            "value": "Luftqualität"
        },
        "refMobilityStation": {
            "type": "Text",
            "value": "urn:ngsi:MobilityStation:KielRegion:Oppendorf"
        },
        "refServiceProvider": {
            "type": "Text",
            "value": "urn:ngsi:AirQualityObserved:Umweltbundesamt:DESH057"
        },
        "service": {
            "type": "Text",
            "value": "/info/airquality"
        }
    },
    {
        "id": "urn:ngsi:MobilityService:KielRegion:Oppendorf:Weather",
        "type": "MobilityService",
        "description": {
            "type": "Text",
            "value": "Infoservice Wetter Bahnhof Oppendorf"
        },
        "location": {
            "type": "geo:json",
            "value": {
                "type": "Point",
                "coordinates": [
                    10.208005,
                    54.326238
                ]
            }
        },
        "name": {
            "type": "Text",
            "value": "Wetter"
        },
        "refMobilityStation": {
            "type": "Text",
            "value": "urn:ngsi:MobilityStation:KielRegion:Oppendorf"
        },
        "refServiceProvider": {
            "type": "Text",
            "value": "urn:ngsi:WeatherObserved:KielRegion:Oppendorf"
        },
        "service": {
            "type": "Text",
            "value": "/info/weather"
        }
    },
    {
        "id": "urn:ngsi:MobilityService:KielRegion:Oppendorf:Parking:Sonderparkplatz",
        "type": "MobilityService",
        "description": {
            "type": "Text",
            "value": "Sonderparkplatz Bahnhof Oppendorf"
        },
        "location": {
            "type": "geo:json",
            "value": {
                "type": "Point",
                "coordinates": [
                    10.208635,
                    54.326435
                ]
            }
        },
        "name": {
            "type": "Text",
            "value": "Sonderparkplatz"
        },
        "refMobilityStation": {
            "type": "Text",
            "value": "urn:ngsi:MobilityStation:KielRegion:Oppendorf"
        },
        "refServiceProvider": {
            "type": "Text",
            "value": "urn:ngsi:OffStreetParking:KielRegion:Oppendorf:SonderParkplatz"
        },
        "service": {
            "type": "Text",
            "value": "/parking/offstreet"
        }
    },
    {
        "id": "urn:ngsi:MobilityService:KielRegion:Hamdorf:Bus",
        "type": "MobilityService",
        "refServiceProvider": {
            "type": "Text",
            "value": "urn:ngsi:TransportStation:AK:05291"
        }
    }
	]

## Die DigiTwin API

Die DigiTwin API ist hier: https://github.com/SmartMaaS-Services/DigiTwin_API/

damit die DigiTwin APP auch ohne Anmeldung funktioniert wurde eine REST API erstellt die den Zugriff auf die Daten der MobilitätsRegion ermöglicht.
Die Folgenden Endpunkte sind in dieser API abgebildet:

Endpunkt | Beschreibung
---------|-------------
/mobility/regions | Die MobilityRegion
/mobility/stations | Die MobilityStations
/mobility/services | Die MobilityServices
/parking/onstreet | Alle Objekte für den Parkservice vom Type OnStreetParking
/parking/offstreet | Alle Objekte für den Parkservice vom Type OffStreetParking
/parking/spots | Alle Objekte für den Parkservice vom Type  ParkingSpot
/transportation/bikehiredockingstations | Alle Objekte für den Transportation Service BikeHire
/transportation/transportstations | Alle Objekte für den Transportation Service Station (Bushaltestellen oder Bahnhöfe)
/info/weather | Alle Objekte für den Info Service Weather
/info/airquality | Alle Objekte für den Info Service Airquality

## Enviroment Variablen der DigiTwin APP

Enviroment | Variablen
---------- | ---------
DIGITWIN_LOG_LEVEL | LogLevel
DIGITWIN_FIWARE_API_URL | Wo findet die App die API
DIGITWIN_FIWARE_UPDATE | Zeit in Sekunden wie oft die API angefragt wird
DIGITWIN_MAP_CENTER_LONGITUDE | LONGITUDE für das Zentrum der MAP
DIGITWIN_MAP_CENTER_LATITUDE | LATITUDE für das Zentrum der MAP 
DIGITWIN_MAP_INITIAL_ZOOM | Zoom Level beim Starten
DIGITWIN_MAP_STATION_ZOOM | Zoom Level für die Staions
DIGITWIN_TRIPGO_API_URL | URL für die Routing API
DIGITWIN_TRIPGO_API_KEY | Key der Routing API
DIGITWIN_TRIPGO_REGION | Region für das Routing


## neue Datenmodelle

Die neuen Datemodelle liegen hier im REPO https://github.com/SmartMaaS-Services/DiGiTwin_DataModel
und sind nach https://smartdatamodels.org/ übertragen worden.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
### Changed
### Deprecated
### Removed
### Fixed
### Security

## [0.0.1]
### Added
- kubernetes deployment added
### Changed
- changed nginx version from 1.19 to 1.21


src/app/services/config.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';

@Injectable({
  providedIn: 'root'
})

export class ConfigService {

  DIGITWIN_LOG_LEVEL: string;

  DIGITWIN_MAP_CENTER_LONGITUDE: number;
  DIGITWIN_MAP_CENTER_LATITUDE: number;
  DIGITWIN_MAP_INITIAL_ZOOM: number;
  DIGITWIN_MAP_STATION_ZOOM: number;

  DIGITWIN_FIWARE_API_URL: string;
  DIGITWIN_FIWARE_UPDATE: number;

  DIGITWIN_TRIPGO_API_URL: string;
  DIGITWIN_TRIPGO_API_KEY: string;
  DIGITWIN_TRIPGO_REGION: string;

  DIGITWIN_NEXTBIKE: string;
  DIGITWIN_GTFS_KVG: string;
  DIGITWIN_GTFS_VKP: string;
  DIGITWIN_GTFS_AK: string;
  DIGITWIN_GTFS_DB: string;

  constructor (private http: HttpClient) {
    this.DIGITWIN_LOG_LEVEL = 'info';

    this.DIGITWIN_FIWARE_API_URL = 'https://api.digitwin-staging.kielregion.addix.io';
    this.DIGITWIN_FIWARE_UPDATE = 10;

    this.DIGITWIN_MAP_CENTER_LONGITUDE = 10.1329672;
    this.DIGITWIN_MAP_CENTER_LATITUDE = 54.3223492;
    this.DIGITWIN_MAP_INITIAL_ZOOM = 12;
    this.DIGITWIN_MAP_STATION_ZOOM = 18;

    this.DIGITWIN_TRIPGO_API_URL = 'https://api.tripgo.com/v1/';
    this.DIGITWIN_TRIPGO_API_KEY = '3102c8594f7223a6fc8a93f2a8b91a57';
    this.DIGITWIN_TRIPGO_REGION = 'DE_SH_Kiel';

    this.DIGITWIN_NEXTBIKE = 'BikeHireDockingStation.json';
    this.DIGITWIN_GTFS_KVG = 'GtfsStopKVG.json';
    this.DIGITWIN_GTFS_VKP = 'GtfsStopVKP.json';
    this.DIGITWIN_GTFS_AK = 'GtfsStopAK.json';
    this.DIGITWIN_GTFS_DB = 'GtfsStopDB.json';
  }

  initialize(): Promise<void> {
    return new Promise<void>((resolve, reject) => {
      let configFile: string = './assets/' + 'config.json';
      this.http.get(configFile).toPromise().then((data: any) => {
        if (this.getConfiguration(data)) {
          resolve();
        } else {
          reject(`Invalid data in file '${configFile}: JSON.stringify(${data})`)
        }
      }).catch((error: any) => {
        reject(`Error reading file '${configFile}: JSON.stringify(${error})`);
      });
    });
  }

  private getConfiguration(config: any): boolean {
    let valid: boolean = true;

    if (config.DIGITWIN_LOG_LEVEL) {
      let logLevels: string[] = ['none', 'info', 'warning', 'error'];
      if (logLevels.includes(config.DIGITWIN_LOG_LEVEL)) {
          this.DIGITWIN_LOG_LEVEL = config.DIGITWIN_LOG_LEVEL;
      } else {
          console.log('DIGITWIN_LOG_LEVEL invalid');
          valid = false;
      }
    } else {
      console.log('DIGITWIN_LOG_LEVEL missing');
      valid = false;
    }

    if (config.DIGITWIN_FIWARE_API_URL) {
      this.DIGITWIN_FIWARE_API_URL = config.DIGITWIN_FIWARE_API_URL;
    } else {
      console.log('DIGITWIN_FIWARE_API_URL missing');
      valid = false;
    }

    if (config.DIGITWIN_FIWARE_UPDATE) {
      this.DIGITWIN_FIWARE_UPDATE = config.DIGITWIN_FIWARE_UPDATE;
    } else {
      console.log('DIGITWIN_FIWARE_UPDATE missing');
      valid = false;
    }

    if (config.DIGITWIN_MAP_CENTER_LONGITUDE) {
      this.DIGITWIN_MAP_CENTER_LONGITUDE = config.DIGITWIN_MAP_CENTER_LONGITUDE;
    } else {
      console.log('DIGITWIN_MAP_CENTER_LONGITUDE missing');
      valid = false;
    }

    if (config.DIGITWIN_MAP_CENTER_LATITUDE) {
      this.DIGITWIN_MAP_CENTER_LATITUDE = config.DIGITWIN_MAP_CENTER_LATITUDE;
    } else {
      console.log('DIGITWIN_MAP_CENTER_LATITUDE missing');
      valid = false;
    }

    if (config.DIGITWIN_MAP_INITIAL_ZOOM) {
      this.DIGITWIN_MAP_INITIAL_ZOOM = config.DIGITWIN_MAP_INITIAL_ZOOM;
    } else {
      console.log('DIGITWIN_MAP_INITIAL_ZOOM missing');
      valid = false;
    }

    if (config.DIGITWIN_MAP_STATION_ZOOM) {
      this.DIGITWIN_MAP_STATION_ZOOM = config.DIGITWIN_MAP_STATION_ZOOM;
    } else {
      console.log('DIGITWIN_MAP_STATION_ZOOM missing');
      valid = false;
    }

    if (config.DIGITWIN_TRIPGO_API_URL) {
      this.DIGITWIN_TRIPGO_API_URL = config.DIGITWIN_TRIPGO_API_URL;
    } else {
      console.log('DIGITWIN_TRIPGO_API_URL missing');
      valid = false;
    }

    if (config.DIGITWIN_TRIPGO_API_KEY) {
      this.DIGITWIN_TRIPGO_API_KEY = config.DIGITWIN_TRIPGO_API_KEY;
    } else {
      console.log('DIGITWIN_TRIPGO_API_KEY missing');
      valid = false;
    }

    if (config.DIGITWIN_TRIPGO_REGION) {
      this.DIGITWIN_TRIPGO_REGION = config.DIGITWIN_TRIPGO_REGION;
    } else {
      console.log('DIGITWIN_TRIPGO_REGION missing');
      valid = false;
    }

    if (config.DIGITWIN_NEXTBIKE) {
      this.DIGITWIN_NEXTBIKE = config.DIGITWIN_NEXTBIKE;
    } else {
      console.log('DIGITWIN_NEXTBIKE missing');
      valid = false;
    }

    if (config.DIGITWIN_GTFS_KVG) {
      this.DIGITWIN_GTFS_KVG = config.DIGITWIN_GTFS_KVG;
    } else {
      console.log('DIGITWIN_GTFS_KVG missing');
      valid = false;
    }

    if (config.DIGITWIN_GTFS_VKP) {
      this.DIGITWIN_GTFS_VKP = config.DIGITWIN_GTFS_VKP;
    } else {
      console.log('DIGITWIN_GTFS_VKP missing');
      valid = false;
    }

    if (config.DIGITWIN_GTFS_AK) {
      this.DIGITWIN_GTFS_AK = config.DIGITWIN_GTFS_AK;
    } else {
      console.log('DIGITWIN_GTFS_AK missing');
      valid = false;
    }

    if (config.DIGITWIN_GTFS_DB) {
      this.DIGITWIN_GTFS_DB = config.DIGITWIN_GTFS_DB;
    } else {
      console.log('DIGITWIN_GTFS_DB missing');
      valid = false;
    }

    return valid;
  }
}
